### DISCUSSION TITLE
New idea: do not limit the usage of roam-block in some specific directories.

### DISCUSSION THREAD

**Discussion by Kinneyzhang** [Kinneyzhang/roam-block#14] [Ideas]
In version 0.0.1, roam-block can only be used in some specific directories defined in ```roam-block-work-home``` variable. The original idea of this limitation is that I want to implement the function what Roam Research have done. The function is that after inputing "(())", it will query database and do completion to insert a block ref. So, roam-block has to cache all blocks of a file when the file is opened or saved. But, I don't want to cache some files which are not necessary to when they are opened or saved, so I set the ``roam-block-work-home``` variable. it consumes that all blocks in these directories' files will be and are necessary to be cached, which will be used to query and insert.

Now, I have got some new ideas! I will not limit the usage of roam-block in the specific directories. For files in global file system, the block in files will be cached only when you store a block-ref or block-embed of it. This means users cannot 'query then insert' for those files but the blocks which have been cached. But I will preserve a variable like ```roam-block-work-home```(the name will be changed) which is used for caching all blocks in those directories. This may be meaningful to be set as the org-roam directory.

BTW, I have mentioned two strategies of inserting a block ref: the first is 'store then insert' (have implemented) and the second id 'query then insert' (will be implemented).

**Comment by nobiot** [comment]
This is great as documenting your design. I did not realise that `roam-block-work-home` was meant to be used to cache all the blocks in the files in the directory — and the cache is to be used for a query-to-insert operation. 

I am really not a technically versed in programming; I use Emacs for my writing and note-taking. I do not use Roam Research, so I cannot comment from that perspective. 

From this perspective, the ability to be able to cache any file outside the directory sounds good. I wonder, though, if it will be confusing a little if some blocks appear in the query, but others do not. Also... if you cache really a large number of blocks from many files, will it slow down the query? — no technical evidence behind this, but just wondered.

**Comment by Kinneyzhang** [comment]
>  if it will be confusing a little if some blocks appear in the query, but others do not.

One should set a variable named  ```roam-block-query-dirs``` and roam-block can only query blocks in dirs defined in this variable. Document it clearly in the variable description should be ok.

> if you cache really a large number of blocks from many files, will it slow down the query? 

One should not worry about the query efficiency of database at a time because database has specially enhancement in querying. But one do has to care about how many times of querying or updating data in database in a second, which is the key point that really affects the performance of roam-block. I have got some new ideas to enhance it, to decrease the frequency of updating database.
